modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java

```diff
@@ -9,145 +9,145 @@
-                if (lastGeneration != null) {
-                    throw addSuppressedExceptions(
-                        new NoSuchFileException(
+                    return new ContentLengthValidatingInputStream(content, start, lastGeneration);
+                } catch (IOException e) {
+                    throw StorageException.translate(e);
+                }
+            } catch (StorageException storageException) {
+                if (storageException.getCode() == RestStatus.NOT_FOUND.getStatus()) {
+                    if (lastGeneration != null) {
+                        throw new NoSuchFileException(
                             "Blob object ["
                                 + blobId.getName()
                                 + "] generation ["
                                 + lastGeneration
                                 + "] unavailable on resume (contents changed, or object deleted): "
                                 + storageException.getMessage()
-                        )
-                    );
-                } else {
-                    throw addSuppressedExceptions(
-                        new NoSuchFileException("Blob object [" + blobId.getName() + "] not found: " + storageException.getMessage())
-                    );
+                        );
+                    } else {
+                        throw new NoSuchFileException("Blob object [" + blobId.getName() + "] not found: " + storageException.getMessage());
+                    }
                 }
-            }
-            if (storageException.getCode() == RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()) {
-                long currentPosition = Math.addExact(start, currentOffset);
-                throw addSuppressedExceptions(
-                    new RequestedRangeNotSatisfiedException(
+                if (storageException.getCode() == RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()) {
+                    throw new RequestedRangeNotSatisfiedException(
                         blobId.getName(),
-                        currentPosition,
-                        (end < Long.MAX_VALUE - 1) ? end - currentPosition + 1 : end,
+                        start,
+                        (end < Long.MAX_VALUE - 1) ? end - start + 1 : end,
                         storageException
-                    )
-                );
+                    );
+                }
+                throw storageException;
             }
-            throw addSuppressedExceptions(storageException);
+        }
+
+        @Override
+        public void onRetryStarted(String action) {
+            // No retry metrics for GCS
+        }
+
+        @Override
+        public void onRetrySucceeded(String action, long numberOfRetries) {
+            // No retry metrics for GCS
+        }
+
+        @Override
+        public long getMeaningfulProgressSize() {
+            return Math.max(1L, GoogleCloudStorageBlobStore.SDK_DEFAULT_CHUNK_SIZE / 100L);
+        }
```

Comment:
The choice of this value is somewhat arbitrary, open to suggestions of whether we should make this consistent across CSPs or use some other value here. SDK default chunk size is 16MB, so this is 160KB